File: python/200.py
xrange = range

class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        n = len(grid)
        if n == 0:
            return 0
        m = len(grid[0])
        for i in xrange(n):
            grid[i] = list(grid[i])
        ans = 0
        for i in xrange(n):
            for j in xrange(m):
                if grid[i][j] == '1':
                    ans += 1
                    grid[i][j] = '0'
                    s = [(i, j)]
                    while s:
                        ui, uj = s.pop()
                        for ai, aj in ((0, 1), (1, 0), (0, -1), (-1, 0)):
                            vi, vj = ui+ai, uj+aj
                            if 0<=vi<n and 0<=vj<m and grid[vi][vj] == '1':
                                grid[vi][vj] = '0'
                                s.append((vi, vj))
        return ans

File: cpp/200.cpp
#include <vector>
#include <utility>
using namespace std;

class Solution {
public:
  int numIslands(vector<vector<char>>& grid) {
    int n = grid.size();
    if(n == 0) return 0;
    int m = grid[0].size();
    int ans = 0;
    vector<int> si, sj;
    vector<int> ai = {0, 0, 1, -1}, aj = {1, -1, 0, 0};
    for(int i=0; i<n; i++)
      for(int j=0; j<m; j++)
        if(grid[i][j] == '1'){
          ans += 1;
          grid[i][j] = '0';
          si.push_back(i); sj.push_back(j);
          while(!si.empty()){
            int ui = si.back(); si.pop_back();
            int uj = sj.back(); sj.pop_back();
            for(int k = 0; k < 4; k++){
              int vi = ui + ai[k], vj = uj + aj[k];
              if(0 <= vi && vi < n && 0 <= vj && vj < m && grid[vi][vj] == '1'){
                grid[vi][vj] = '0';
                si.push_back(vi); sj.push_back(vj);
              }
            }
          }
        }
    return ans;
  }
};
